Remove RaCandidate projector in favor of query

pablothedude opened the pull request:
This is done in order to mitigate performance issues when updating the institution configuration.
If the configuration was updated all ra candidates for select_raa institutions needed to be checked
and updated if required. Which resulted in timouts and OOM's.

Initially a view was used to prevent updateing the RACandidate through events but this proved
slow when filtering was done on the view in some cases. Therefore the view was removed and
the query itself was moved to the repository so filtering could be handled earlier on which
greatly  improved performance.

https://www.pivotaltracker.com/story/show/172245466
https://github.com/OpenConext/Stepup-Deploy/pull/105

MKodde commented on README.md:
```diff
@@ -33,7 +33,7 @@ Some of the configuratio of the components is static (i.e. stored in parameteres
 Whenever adding a new event, be sure to update `app/config/events.yml`.
 This is a list of events that is shown when replaying events.
-Also be sure to create or update the event serialization/deserialization tests,
+Also be sure to create or update the event serialization/deserialization tests,RaCandidateRepository
```
Accidental pasted RaCandidateRepository?

MKodde commented on app/DoctrineMigrations/Version20200416135127.php:
```diff
@@ -0,0 +1,49 @@
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('CREATE VIEW `view_ra_candidate` AS
+(
+SELECT DISTINCT i.id          as identity_id,
+                i.institution,
+                i.common_name,
+                i.email,
+                i.name_id,
+                a.institution AS ra_institution
+FROM identity i
+         INNER JOIN vetted_second_factor vsf on vsf.identity_id = i.id
+         INNER JOIN institution_authorization a ON (
+        a.institution_role = \'select_raa\' AND a.institution_relation = i.institution
+    )
+WHERE NOT EXISTS(SELECT 1 FROM ra_listing AS l WHERE l.identity_id = i.id AND l.ra_institution = a.institution)
+);');
```
3 nitpicks:
1. For the sake of readability consider using HEREDOC to capture the query.
2. My inner SQL linter triggered on mixed-case use of MySQL keywords (`AS` and `as`, `on`). You also not use them consistently. For the aliases you create for the table names (sh|c)ould also have an `AS` keyword.
3. Consider changing the where not exists subquery into a multi-liner.

pablothedude replied:
Done

MKodde commented on app/DoctrineMigrations/Version20200416135127.php:
```diff
@@ -0,0 +1,49 @@
+        // this down() migration is auto-generated, please modify it to your needs
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('DROP VIEW `view_ra_candidate` ');
+        $this->addSql('DROP TABLE ra_candidate');
```
I'm confused why the ra_candidate table is also dropped when the migration is downed. Would expect it to be dropped on the UP method. Could you explain this maybe also in an in-line comment?

pablothedude replied:
Thanks, that indeed shouldn't happen at all.

pablothedude replied:
Removed it completely. I'll consult our PM what should happen with it. Maybe we could better clean it the next release.

pablothedude replied:
The RA candidate could be removed so I added the drop statement in the `up` method and added a irreversible exception to the `down` method.

MKodde commented on src/Surfnet/StepupMiddleware/ApiBundle/EventListener/IgnoreTablesListener.php:
```diff
@@ -0,0 +1,70 @@
+        foreach ($tableNames as $tableName) {
+            if (in_array($tableName, $this->ignoredTables)) {
+                // remove table from schema
+                $schema->dropTable($tableName);
```
What a confusing methodName for that action. It implies dropping the view/table from the database schema. Where it actually stops tracking it. Correct me if I'm wrong. And yes, I know this is Doctrine ORM, not our code.

pablothedude replied:
I don't like the listener either but that seemed to be the only solution. Initially the view is added so it needs to be dropped to let the schema tool work with the view.

MKodde commented on src/Surfnet/StepupMiddleware/ApiBundle/Identity/Entity/RaCandidate.php:
```diff
@@ -27,39 +27,24 @@
- *          @ORM\UniqueConstraint(name="idx_ra_candidate_unique_identity_institution", columns={"identity_id", "ra_institution"})
- *     }
- * )
+ * Be aware that this entity is used for the RA Candidate view. This entity shouldn't be used to store any RA candidates.
+ * The IgnoreTablesListener is used to prevent schema changes when the Doctrine Schema tool is used.
```
:+1:

MKodde commented on src/Surfnet/StepupMiddleware/ApiBundle/Identity/Entity/RaCandidate.php:
```diff
@@ -94,6 +79,10 @@ class RaCandidate implements JsonSerializable
      */
     public $email;
 
+    private function __construct()
+    {
+    }
```
Was adding a private constructor important? And why?

pablothedude replied:
This is to prevent instantiating an Entity which shouldn't be initiated at all because a view is used.

pablothedude commented:
@MKodde I've removed the view and moved the query used to the RaCandidate repository. This greatly optimized the speed to return a result because the DBMS could optimize the query much better.

The filtering on the ra candidate page is changed a bit because it will show all institutions even if there are no possible ra-candidates. This won't outweigh the benefits of the improved performance because the only drawback would be an empty resultset when selected.

I've done the change in a separate commit so it would become easier to review.

MKodde commented on src/Surfnet/StepupMiddleware/ApiBundle/Identity/Repository/RaCandidateRepository.php:
```diff
@@ -67,42 +86,42 @@ public function createSearchQuery(RaCandidateQuery $query)
         if (!empty($query->secondFactorTypes)) {
             $queryBuilder
-                ->innerJoin(VettedSecondFactor::class, 'vsf', Join::WITH, 'rac.identityId = vsf.identityId')
+                //->innerJoin(VettedSecondFactor::class, 'vsf', Join::WITH, 'rac.identityId = vsf.identityId')
```
Perhaps permanently remove this?

MKodde commented on app/DoctrineMigrations/Version20200416135127.php:
```diff
@@ -0,0 +1,30 @@
+    {
+        $this->throwIrreversibleMigrationException('This migration is irreversible and cannot be reverted because it will need a replay on the RACandidateProjector.');
+
+        //$this->addSql('DROP INDEX idx_vetted_second_factor_type ON vetted_second_factor');
```
Remove this permanently? If you want to keep it for future reference, then please add a comment.

pablothedude replied:
Removed it completely

MKodde commented on src/Surfnet/StepupMiddleware/ApiBundle/Identity/Entity/RaCandidate.php:
```diff
@@ -27,39 +27,24 @@
+ * Be aware that this entity is used for the RA Candidate view. This entity shouldn't be used to store any RA candidates.
+ * The IgnoreTablesListener is used to prevent schema changes when the Doctrine Schema tool is used.
+ *
+ * @ORM\Table(name="view_ra_candidate")
```
This might need revision after venturing away from the view?

pablothedude replied:
Yes, that's true and should have been removed. I've directly removed the `IgnoreTablesListener` and the creation of the view in Deploy which could then also be removed.

MKodde commented:
I've functionally tested this after removing the view. I did not find any functional regressions. The ra candidate listing seemed to remain identical between commits. :+1:
